Support NCCL Premul Sum

crcrpar opened the pull request:
This PR adds the support for https://docs.nvidia.com/deeplearning/nccl/archives/nccl_21212/user-guide/docs/api/ops.html?highlight=premul#c.ncclRedOpCreatePreMulSum.

The major changes include
- convert enum ReduceOp to struct
- add premul sum specific paths to init.cpp and Ops.cpp.

note:
- For pip wheels / conda binaries to support this, ~~I think https://github.com/pytorch/pytorch/pull/79132 would be needed~~ https://github.com/pytorch/pytorch/pull/82775 landed

The commit titled "add nccl premul" whose current hash is https://github.com/pytorch/pytorch/pull/81272/commits/cb99ad67447b5899ecf8c4c3d78deaafa1cc09b8 was authored by @mcarilli and @ptrblck.

cc @ptrblck

ptrblck committed:
add nccl premul

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>
Co-authored-by: ptrblck <ptrblck@users.noreply.github.com>
Co-authored-by: Michael Carilli <mcarilli@nvidia.com>

Patch for https://github.com/pytorch/pytorch/pull/79582

Apparently 79852 is newer than 34 and the commit below so the PR assumes
`ReduceOp` to be an `enum`, not a `struct` including an `enum` inside
it.

crcrpar committed:
implcit conversion of ::c10d::ReduceOp and ::c10d::ReduceOp::Kind

Build looks okay.

But the test case of PREMUL_SUM is failing

```
======================================================================
ERROR: test_allreduce_ops (__main__.ProcessGroupNCCLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File
"/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py",
line 501, in wrapper
    self._join_processes(fn)
  File
"/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py",
line 724, in _join_processes
    self._check_return_codes(elapsed_time)
  File
"/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py",
line 769, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File
"/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py",
line 622, in run_test
    getattr(self, test_name)()
  File
"/opt/conda/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py",
line 503, in wrapper
    fn()
  File "/opt/pytorch/pytorch/test/distributed/test_c10d_nccl.py", line
350, in test_allreduce_ops
    allreduce(tensors, c10d.make_nccl_premul_sum(factor))
  File "/opt/pytorch/pytorch/test/distributed/test_c10d_nccl.py", line
317, in allreduce
    work = pg.allreduce(tensors, opts)
RuntimeError: op_ != PREMUL_SUM INTERNAL ASSERT FAILED at
"/opt/pytorch/pytorch/torch/csrc/distributed/c10d/Types.hpp":44, please
report a bug to PyTorch. PREMUL_SUM requires a scale factor tensor or
scalar argument
Exception raised from ReduceOp at
/opt/pytorch/pytorch/torch/csrc/distributed/c10d/Types.hpp:44 (most
recent call first):
```

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

Try&Error 1. Pass ReduceOp in Ops.cpp

With this change, I'm seeing

```
In file included from
/home/mkozuki/ghq/github.com/crcrpar/pytorch/torch/csrc/distributed/c10d/Ops.cpp:1:
In file included from
/home/mkozuki/ghq/github.com/crcrpar/pytorch/torch/csrc/distributed/c10d/Ops.hpp:4:
In file included from
/home/mkozuki/ghq/github.com/crcrpar/pytorch/torch/csrc/distributed/c10d/ProcessGroup.hpp:10:
In file included from
/home/mkozuki/ghq/github.com/crcrpar/pytorch/aten/src/ATen/ATen.h:18:
In file included from
/home/mkozuki/ghq/github.com/crcrpar/pytorch/aten/src/ATen/TensorIndexing.h:14:
In file included from
/home/mkozuki/ghq/github.com/crcrpar/pytorch/aten/src/ATen/core/List.h:483:
In file included from
/home/mkozuki/ghq/github.com/crcrpar/pytorch/aten/src/ATen/core/List_inl.h:4:
In file included from
/home/mkozuki/ghq/github.com/crcrpar/pytorch/aten/src/ATen/core/ivalue.h:1369:
/home/mkozuki/ghq/github.com/crcrpar/pytorch/aten/src/ATen/core/ivalue_inl.h:1656:59:
error: no type named 'element_type' in 'c10d::ReduceOp'
  using ElemType = typename std::remove_pointer<T>::type::element_type;
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
```

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

Revert "Try&Error 1. Pass ReduceOp in Ops.cpp"

This reverts commit 2e15955eab248f7a8d81b97c2ef653e4a50d8577.

Add C++ only constructor to ReduceOp to supply supplement

Build is okay

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

cosmetic

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Define NCCL PREMUL_SUM as different ops

to dodge the limitation of torch jit script which disables exposing
`struct ReduceOp`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
rely on implicit conversion c10d.ReduceOp.Kind -> c10d.ReduceOp

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
declaration in header and knob in init.cpp

Remove redundant ReduceOp handling

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

cosmetic

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

tentatively comment out version guard macro

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

fix build but outputs are zeroed...

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

get more dependent on implicit conversion

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Choose the correct path according to type of scaling factor

oh boy, you should have been confused with the condition of scalar
scaling factor or tensor scaling factor...

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

remove unnecessary ReduceOp::Kind

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
type cast to uint64_t from ReduceOp::Kind

crcrpar committed:
Declare `AutoNcclGroup` in torch/csrc/cuda/nccl.h

and define it in torch/csrc/cuda/nccl.cpp so that
torch/csrc/distributed/c10d/NCCLUtils.cpp can remove the identical
definition.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
give version guards.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
build only when USE_NCCL

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
better version guard

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
missing `ENABLE_NCCL_PREMUM_SUM_SUPPORT`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
lintrunner

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
skip if NCCL does not support premul sum

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
handle PREMUL_SUM in Gloo

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
handle PREMUL_SUM

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
lintrunner

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

flake8

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

clangformat

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

cosmetic fix

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

aazzolini commented:
@kwen2501 could you review this?

crcrpar committed:
Remove wrong version guard.

These files cannot see USE_NCCL nor USE_C10D_NCCL

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

Move version guard

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

lintrunner - clang-format

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

aazzolini commented:
Tagging @kwen2501 -- could you follow up with a review when appropriate?

kwen2501 commented:
Hi @crcrpar @aazzolini thanks for the PR. I am reviewing now.

kumpera commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1018,6 +1051,16 @@ that adds a prefix to each key inserted to the store.
                  const std::vector<at::Tensor>& tensors,
                  const ::c10d::AllreduceOptions& opts) {
+                if (opts.reduceOp.op_ == ::c10d::ReduceOp::PREMUL_SUM) {
+#if defined(ENABLE_NCCL_PREMUL_SUM_SUPPORT)
```
We should have this sort of hacks in the backend agnostic part of the code.
It should be up to each backend to detect and fail if the user tries an unsupported reducion

crcrpar replied:
This is the only way I could find to deal with the requirement of `::c10d::ops:allreduce`, assuming `struct ReduceOp` having a scaling factor is a correct way to go.

kwen2501 replied:
@kumpera Do you mean we should *not* have this sort of hacks in the backend agnostic part of the code?

kwen2501 replied:
I agree that once the dispatcher->implementation route is in place, this kind of check can be put into the CUDA/NCCL specific implementation, rather than for the general ProcessGroup part.

kumpera commented on torch/csrc/distributed/c10d/Types.hpp:
```diff
@@ -3,20 +3,92 @@
+  // Right now, only PREMUL_SUM needs supplementary data, but the same
+  // mechanism could extend to support other nontrivial reduce ops with
+  // different supplementary payloads.
+  std::shared_ptr<_SupplementBase> supplement_;
```
This should be intrusive_ptr, no reason to use std::shared_ptr.

crcrpar replied:
the contract of having `refcount_` inside Supplement is unclear to me.

kumpera commented on torch/csrc/distributed/c10d/Types.hpp:
```diff
@@ -3,20 +3,92 @@
+  std::shared_ptr<_SupplementBase> supplement_;
+};
+
+template<typename T> ReduceOp makeNCCLPreMulSum(const T& factor) {
```
This can be in init.cpp

crcrpar replied:
then ops.cpp needs to depend on init.cpp so the current position looks better.

kumpera commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -27,7 +28,41 @@ c10::intrusive_ptr<ProcessGroup::Work> allreduce_(
+// NCCL specific PREMUL_SUM Reductions
+// NCCL 2.11.1-1 introduced "User Defined Reduction Operators":
+//   https://docs.nvidia.com/deeplearning/nccl/archives/nccl_21212/user-guide/docs/api/ops.html
+c10::intrusive_ptr<ProcessGroup::Work>
```
Are the two new PG methods required or just out of convenience for users?

Introducing those methods in the backend agnostic PG class is not a good idea.

They can be exposed as helpers in distributed_c10.py that do the same.

kwen2501 commented on torch/_C/_distributed_c10d.pyi:
```diff
@@ -408,3 +413,4 @@ def _verify_params_across_processes(
     params: List[Tensor],
     logger: Optional[Logger],
 ): ...
+def make_nccl_premul_sum(factor: Union[float, List[Tensor]]) -> ReduceOp: ...
```
Can we underscore this function name for now? Because there is a chance this function be shuffled or name changed later.

crcrpar committed:
private make_nccl_premul_sum

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
uh oh, I noticed that I need to update the other `c10d::ops` that interact with `ReduceOp` such as `reduce_scatter`

crcrpar committed:
Handling PREMUL_SUM for ops which use `c10d::ReduceOp`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
fix some signatures and add test

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
uh, looks like I need some glue so that we can compare `reduce_op: c10d.ReduceOp` against `reduce_op_kind: c10d.ReduceOp.Kind` in python

crcrpar committed:
Define `__eq__` and `__hash__` for `c10d.ReduceOp`

These had been available as `::c10d::ReduceOp` had been an enum.

One use-case is
https://github.com/pytorch/xla/blob/7b6747dfbe10d43fe7fdc6873f64f793fd714a8c/torch_xla/distributed/xla_backend.py#L48

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

clang-format

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

kwen2501 commented on torch/csrc/distributed/c10d/Types.hpp:
```diff
@@ -71,6 +71,14 @@ struct ReduceOp {
     return other == op_;
   }
 
+  bool operator==(const ReduceOp::Kind other) {
```
nit: not sure if `Kind` conveys the clearest meaning when a reader reads it somewhere else than where it is defined.
Would `RedOpType` make it clearer? I thought about `OpType` to make it shorter, but it seems `OpType` has been taken in `ProcessGroupNCCL`.

kwen2501 commented:
Hi @crcrpar thanks for the updates. 
I guess the main comment from @kumpera and me is to try to avoid adding `nccl_premulsum_allreduce` in Ops.hpp/cpp if possible. It appears to me that you have already broadened the definition of reduce op, can that structure and `AllreduceOption` encapsulate the PreMul information and pass it down to ProcessGroupNCCL?

crcrpar commented:
If I understand things correctly, to avoid `nccl_premulsum_allreduce` in Ops.hpp/cpp needs some modifications in Torch Script so that Torch Script compiler supports `AllreduceOption` and / or `ReduceOp`.

crcrpar committed:
ReduceOp::Kind -> ReduceOp::RedOpType

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

kwen2501 reviewed (commented):
The Kind -> RedOpType change looks good to me! Thanks!

crcrpar commented:
@kwen2501 what's your take on Ops.cpp? I don't think there's other ways to avoid NCCL specific ops in Ops.cpp other than to update Torch Script.

crcrpar committed:
update `getNcclReduceOp` call in _reduce_op

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
lintrunner - clang-format

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
@kwen2501 how can we move this forward?

crcrpar committed:
Remove NCCL PREMUL_SUM specifics from Ops.cpp/.hpp

by giving up having `reduce` ops with nccl premul_sum conforming to
TorchScript Compiler requirements.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

kwen2501 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1018,6 +1068,16 @@ that adds a prefix to each key inserted to the store.
+#if defined(ENABLE_NCCL_PREMUL_SUM_SUPPORT)
+                  if (self->getBackendName() == "nccl") {
+                    std::vector<at::Tensor> non_const_tensors = tensors;
+                    return self->allreduce(non_const_tensors, opts);
```
Essentially, you are bypassing the dispatcher by calling `self->allreduce` here. It would lead to an escape of traceability by the PyTorch compiler.
Temporarily it might be okay, but in long term we are going to require all ops go through the dispatcher.

kwen2501 replied:
Regarding your question of how the dispatcher can recognize custom data types, I wonder if this reference code could help:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/Ops.cpp#L142-L145

crcrpar replied:
yeah I know this is dodging the effort of making c10d ops traceable. 
this is my "workaround" to keep "Ops" clean due to my lack of knowledge on TorchScript.
let's access this in a follow-up PR

kwen2501 reviewed (approved):
Generally look good to me. I added some comments.
If you could address some or all of them in this PR, that would be awesome. Otherwise, we can address them (particularly the dispatcher support) in a follow-up PR if this feature is of urgent need.

crcrpar committed:
remove redundant check and record

as per [Ke's comment](https://github.com/pytorch/pytorch/pull/81272/files#r952251249)

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1772,19 +1807,36 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::pointToPoint(
+      tensor.scalar_type(), // dType
+      std::vector<int64_t>(), // inSplitSizes
+      std::vector<int64_t>()); // outSplitSizes
+
```
no need to do I believe thus removed in https://github.com/pytorch/pytorch/pull/81272/commits/86856488daf90d5cd6c362ebc1afb8c62e55f676

crcrpar committed:
add todo about ReduceOp

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
@pytorchmergebot merge

crcrpar commented:
The failure [here](https://github.com/pytorch/pytorch/runs/7981181277?check_suite_focus=true) doesn't look related to this PR...

```console
2022-08-23T19:25:21.8641906Z   File "test/run_test.py", line 990, in <module>
2022-08-23T19:25:21.8650908Z     main()
2022-08-23T19:25:21.8651130Z   File "test/run_test.py", line 968, in main
2022-08-23T19:25:21.8651362Z     raise RuntimeError(err_message)
2022-08-23T19:25:21.8651596Z RuntimeError: test_nn failed! Received signal: SIGKILL
2022-08-23T19:25:22.5999254Z 
2022-08-23T19:25:22.5999671Z real	2m50.787s
2022-08-23T19:25:22.5999943Z user	6m41.756s
2022-08-23T19:25:22.6000151Z sys	0m54.374s
2022-08-23T19:25:22.8195029Z ##[error]Process completed with exit code 1.
```

kwen2501 commented:
Just try rerun the failed test or rebase if master is clean

crcrpar commented:
@pytorchmergebot merge

weiwangmeta commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

weiwangmeta commented:
Hi @crcrpar  Thank you for your contribution, however this PR has failed our internal testing. Some more details would be provided later on upon successful revert.

weiwangmeta commented:
This PR has introduced a new switch option which seems to have confused some of internal testing. We might be able to re-open and re-merge.
